sdk/core/core/test/cmocka/test_az_url_encode.c

```diff
@@ -44,29 +47,168 @@ static uint8_t url_decoded_buf[] = {
   0xF0, 0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB, 0xFC, 0xFD, 0xFE, 0xFF
 };
 
-static void test_url_encode(void** state)
+static void test_url_encode_basic(void** state)
 {
   (void)state;
   uint8_t buf[256 * 3] = { 0 };
   az_span const buffer = AZ_SPAN_FROM_BUFFER(buf);
   int32_t url_length = 0;
 
-  assert_true(az_succeeded(
-      _az_span_url_encode(buffer, AZ_SPAN_FROM_STR("https://vault.azure.net"), &url_length)));
+  {
+    // Empty
+    uint8_t buf0[0] = {};
+    az_span const buffer0 = AZ_SPAN_FROM_BUFFER(buf0);
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer0, AZ_SPAN_NULL, &url_length)));
+    assert_true(az_span_is_content_equal(az_span_slice(buffer0, 0, url_length), AZ_SPAN_NULL));
+    assert_int_equal(url_length, 0);
+  }
+  {
+    // Just enough to do no encoding
+    uint8_t buf5[5] = { 0 };
+    az_span const buffer5 = AZ_SPAN_FROM_BUFFER(buf5);
+
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer5, AZ_SPAN_FROM_STR("hello"), &url_length)));
+
+    assert_int_equal(url_length, sizeof("hello"));
+    assert_true(az_span_is_content_equal(buffer5, AZ_SPAN_FROM_STR("hello")));
+
+    // Insufficient size
+    buf5 = { '*', '*', '*', '*', '*' };
+    az_span const buffer2 = az_span_slice(buffer5, 0, 2);
+    url_length = 0xFF;
+
+    assert_true(
+        _az_span_url_encode(buffer2, AZ_SPAN_FROM_STR("/"), &url_length)
+        == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
+
+    assert_int_equal(url_length, 0);
+    assert_true(az_span_is_content_equal(buffer5, AZ_SPAN_FROM_STR("*****")));
+
+    // Sufficient size
+    az_span const buffer3 = az_span_slice(buffer5, 0, 3);
+    url_length = 0xFF;
+
+    assert_true(az_succeeded(_az_span_url_encode(buffer3, AZ_SPAN_FROM_STR("/"), &url_length)));
+    assert_int_equal(url_length, 3);
+    assert_true(az_span_is_content_equal(buffer5, AZ_SPAN_FROM_STR("%2F**")));
+  }
+  {
+    // May be sufficient to perform encoding
+    uint8_t buf7[7] = { '*', '*', '*', '*', '*', '*', '*' };
+    az_span const buffer7 = AZ_SPAN_FROM_BUFFER(buf7);
+    az_span const buffer5 = az_span_slice(buffer7, 0, 5);
+    az_span const buffer6 = az_span_slice(buffer7, 0, 6);
+    {
+      url_length = 0xFF;
+      assert_true(
+          _az_span_url_encode(buffer5, AZ_SPAN_FROM_STR("he/lo"), &url_length)
```

Comment:
nit: store these in named locals and reuse (place around line 99-102 in the containing scope)

```C
az_span input = AZ_SPAN_FROM_STR("he/lo");
az_span expected = AZ_SPAN_FROM_STR("*****");
```

Reply:
I reworked these buffers and I think I added more clarity. Instead of reusing, I declare everything inside a very small scope, that is sufficient just for one scenario, I think it is better now.